core/trino-main/src/main/java/io/trino/server/security/oauth2/NimbusHttpClient.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.server.security.oauth2;
+
+import com.google.common.collect.ImmutableMultimap;
+import com.nimbusds.oauth2.sdk.http.HTTPRequest;
+import io.airlift.http.client.HttpClient;
+import io.airlift.http.client.Request;
+import io.airlift.http.client.ResponseHandler;
+
+import javax.inject.Inject;
+
+import java.net.URI;
+
+import static com.nimbusds.oauth2.sdk.http.HTTPRequest.Method.DELETE;
+import static com.nimbusds.oauth2.sdk.http.HTTPRequest.Method.GET;
+import static com.nimbusds.oauth2.sdk.http.HTTPRequest.Method.POST;
+import static com.nimbusds.oauth2.sdk.http.HTTPRequest.Method.PUT;
+import static io.airlift.http.client.StaticBodyGenerator.createStaticBodyGenerator;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class NimbusHttpClient
```

Comment:
I've simplified it a bit further but I couldn't use a single since it was causing circular dependency:
`IdpClient -> OAuth2Endpoints -> IdpClient` as `OAuth2Endpoints` is both: a dependency and a product of `IdpClient`.

Reply:
Ok, so I think we can be a lot more concise and remove a lot of unnecessary code here, while also cleaning up the separation between `OAuth2Service` and it's `OAuth2Client`. At a high level I propose we expand the scope of `OAuth2Client` (i.e. `NimbusClient`), reduce the scope of `OAuth2Service`, which will eliminate the configuration complexity by moving it to an init method on `NimbusClient` rather than trying to build it within Guice.

Specifically, `OAuth2Client` will gain the responsibilities of resolving access tokens and calling the `/userinfo` endpoint if requested; this will include validating the ID and access tokens. Correspondingly `OAuth2Service` will lose those responsibilities, which it really shouldn't have in the first place. 

Additionally I propose that we eliminate the complexity around building configuration by keeping the conditional module, but having `OidcDiscoveryModule` and `OAuth2StaticConfigurationModule` only bind their respective configuration objects. We then do an `Optional` injection of both config objects into `NimbusClient` and register an initialization method either as per https://github.com/trinodb/trino/pull/9788#discussion_r755966135 or with the `@PostConstruct` annotation (@kokosing can you tell me what the difference between those two methods is?). The initialization method will then either pull the values from the static config or dynamically retrieve them, as necessary. 

In conclusion, this should remove a large part of the (shared) configuration from `OAuth2Service` and make `NimbusClient` the sole consumer of anything that would be present in the OIDC metadata eliminating the need to dynamically generate that configuration, while reducing the number of objects to be juggled and clarifying the interface between the service and it's client.